rust/lance-index/src/scalar/bitmap.rs

```diff
@@ -0,0 +1,283 @@
+
+        Ok(Arc::new(Self::try_from_serialized(
+            serialized_lookup,
+            store,
+        )?))
+    }
+
+    /// Remap the row ids, creating a new remapped version of this index in `dest_store`
+    async fn remap(
+        &self,
+        mapping: &HashMap<u64, Option<u64>>,
+        dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+
+        unimplemented!()
+    }
+
+    /// Add the new data into the index, creating an updated version of the index in `dest_store`
+    async fn update(
+        &self,
+        new_data: SendableRecordBatchStream,
+        dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+
+        unimplemented!()
+    }
+}
+
+fn get_batch_from_arrays(keys: Arc<dyn Array>, binary_bitmaps: Arc<dyn Array>, pages: Arc<dyn Array>) -> Result<RecordBatch> 
+{
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("keys", keys.data_type().clone(), true),
+        Field::new("bitmaps", binary_bitmaps.data_type().clone(), true),
+        Field::new("pages", pages.data_type().clone(), true),
+    ]));
+
+    let columns = vec![
+        keys,
+        binary_bitmaps,
+        pages
+    ];
+
+    Ok(RecordBatch::try_new(schema, columns)?)
+}
+
+// Takes an iterator of Vec<u64> and processes each vector
+// to turn it into a RoaringTreemap. Each RoaringTreeMap is
+// serialized to bytes. The entire collection is converted to a BinaryArray
+fn get_bitmaps_from_iter<I>(iter: I) -> Arc<dyn Array> 
+where 
+    I: Iterator<Item = Vec<u64>>
+{
+    let mut builder = BinaryBuilder::new();
+    iter.for_each(|vec| {
+        let mut bitmap = RoaringTreemap::new();
+        bitmap.extend(vec.into_iter());
+        let mut bytes = Vec::new();
+        bitmap.serialize_into(&mut bytes).unwrap();
+        builder.append_value(&bytes);
+    });
+
+    Arc::new(builder.finish())
+}
+
+// Takes an iterator of HashSet<u64> and processes each set
+// to turn it into an Array UInt64Array. The entire collection is stored 
+// as a variable length ListArray
+fn get_pages_from_iter<I>(iter: I) -> Arc<dyn Array>
+where
+    I: Iterator<Item = HashSet<u64>>
+{
+    let mut builder = ListBuilder::new(UInt64Builder::new());
+    iter.for_each(|set| {
+        let pages_vec: Vec<u64> = set.into_iter().collect();
+        builder.values().append_slice(&pages_vec);
+        builder.append(true);
+    });
+
+    let pages_array = builder.finish();
+
+    Arc::new(pages_array)
+}
+
+pub async fn train_bitmap_index(
+    data_source: Box<dyn BtreeTrainingSource + Send>,
+    index_store: &dyn IndexStore,
+) -> Result<()> {
+    let mut batch_idx = 0;
+    let mut batches_source = data_source.scan_ordered_chunks(4096).await?;
+
+    // mapping from item to list of row ids where it is present
+    let mut dictionary: HashMap<ScalarValue, Vec<u64>> = HashMap::new();
+    // mapping from item to set of pages where it is present
+    let mut page_dictionary: HashMap<ScalarValue, HashSet<u64>> = HashMap::new();
```

Comment:
Hmm I'm not sure the page dictionary will make sense in the future. We'd like to re-use indices after compaction. That is, two data files might be merged together, but the rows keep the same row ids (while the page structure might change). That's what the [move-stable row id epic](https://github.com/lancedb/lance/issues/2307) is working towards. So having values that are tightly coupled to the particular data file may not make sense in the future. cc @westonpace

Reply:
Yeah, I removed the page dictionary. We can directly use the bitmaps